lib/bolt/cli.rb

```diff
@@ -564,19 +564,122 @@ def generate_types
+      # dependencies are caught early and do not create a project directory.
+      if options[:modules]
+        if puppetfile.exist?
+          raise Bolt::CLIError,
+                "Found existing Puppetfile at #{puppetfile}, unable to initialize project with "\
+                "module#{plural} #{options[:modules].join(', ')}"
+        else
+          puppetfile_specs = resolve_puppetfile_specs
+        end
+      end
 
-      result = if ok
-                 "Successfully created Bolt project directory at #{path}"
-               else
-                 "Could not create Bolt project directory at #{path}"
-               end
-      outputter.print_message result
+      # Warn the user if the project directory already exists. We don't error here since users
+      # might not have installed any modules yet.
+      if config.exist?
+        @logger.warn "Found existing project directory at #{boltdir}"
+      end
 
-      ok ? 0 : 1
+      # Create the project directory
+      FileUtils.mkdir_p(boltdir)
+
+      # Bless the project directory as a boltdir
+      if FileUtils.touch(config)
+        outputter.print_message "Successfully created Bolt project at #{boltdir}"
+      else
+        raise Bolt::FileError.new("Could not create Bolt project directory at #{boltdir}", nil)
+      end
+
+      # Write the generated Puppetfile to the fancy new boltdir
+      if puppetfile_specs
+        if File.write(puppetfile, puppetfile_specs.join("\n"))
+          outputter.print_message "Successfully created Puppetfile at #{puppetfile}"
+          # Install the modules from our shiny new Puppetfile
+          if install_puppetfile({}, puppetfile, modulepath)
+            outputter.print_message "Successfully installed module#{plural} #{options[:modules].join(', ')}"
+          else
+            raise Bolt::CLIError, "Could not install module#{plural} #{options[:modules].join(', ')}"
+          end
+        else
+          raise Bolt::FileError.new("Could not write Puppetfile at #{puppetfile}", nil)
+        end
+      end
+
+      0
+    end
+
+    # Resolves Puppetfile specs from user-specified modules and dependencies resolved
+    # by the puppetfile-resolver gem.
+    def resolve_puppetfile_specs
+      require 'puppetfile-resolver'
+
+      # Convert module names to module specifications. The Puppetfile resolver requires
+      # that a version be specified, so we'll default to the latest.
+      modules = options[:modules].map do |mod|
+        "mod '#{mod}', :latest"
+      end
+
+      # Builds a model of the Puppetfile content that is used to build the dependency graph
+      model = PuppetfileResolver::Puppetfile::Parser::R10KEval.parse(modules.join("\n"))
+
+      # Make sure the Puppetfile model is valid
+      unless model.valid?
+        plural = options[:modules].count == 1 ? '' : 's'
+        raise Bolt::ValidationError,
+              "Unable to resolve dependencies for module#{plural} #{options[:modules].join(', ')}"
+      end
```

Comment:
@beechtom You don't really need any of the parsing stuff.  The parser is designed to take a text string and turn it into a document model.  However, you already know the module names and versions, so you can construct the model automatically.

e.g.

```
model = PuppetfileResolver::Puppetfile::Document.new('')
options[:modules].each do |mod_name|
  model.add_module(
    PuppetfileResolver::Puppetfile::ForgeModule.new(mod_name).tap { |mod| mod.version = :latest }
)
end
```

Reply:
It's one of the main reasons why I abstracted out the model, so you can do this kinda stuff.

Reply:
Fantastic, thanks! This is _much_ better than creating a string.